neurodamus/metype.py

```diff
@@ -274,6 +275,69 @@ def local_to_global_coord_mapping(self, points):
     def delete_axon(self):
         self._cellref.replace_axon()
 
+    def get_all_segment_points(self):
+        """This method retrieves the extreme points of every neuron segment,
+        and assign to attribute self.all_segment_points
+        Returns: {section_name: list of the [x,y,z] coordinates of the segment end points}
+        e.g. a soma with 3 segments -> {"cell.soma[0]": list 4 [x,y,z] points}
+        """
+
+        def get_seg_extremes(sec, loc2glob):
+            """Get extremes and roto-translate in global coordinates
+            Args:
+                sec: the hoc section
+                loc2glob: matrix to convert from local to global coordinates
+            """
+
+            def get_local_seg_extremes(nseg, pp):
+                """Compute the position of beginning and end of each compartment in a section
+
+                Assumption: all the compartments have the same length
+
+                Args:
+                    nseg: number of compartments.
+                    pp: a nX4 matrix of positions of points.
+                    The first col give the relative position, (x in neuron) along the
+                    axis. It is in the interval [0, 1].
+                    The other 3 columns give x,y,z triplets of the points in a global system of
+                    reference.
+
+                Returns:
+                - a matrix 3Xn of the position of the extremes of every proper compartment
+                """
+                if not pp:
+                    # when section has no 3d points, i.e. new axons, return placeholder
+                    return np.array([])
+                pp = np.array(pp)
+                x_rel, xp, yp, zp = pp[:, 0], pp[:, 1], pp[:, 2], pp[:, 3]
+                x = np.linspace(0, 1, nseg + 1)
+                xp_seg = np.interp(x, x_rel, xp)
+                yp_seg = np.interp(x, x_rel, yp)
+                zp_seg = np.interp(x, x_rel, zp)
+
+                return np.transpose([xp_seg, yp_seg, zp_seg])
+
+            ll = [
+                [sec.arc3d(i) / sec.L, sec.x3d(i), sec.y3d(i), sec.z3d(i)] for i in range(sec.n3d())
+            ]
+            ans = get_local_seg_extremes(
+                sec.nseg,
+                ll,
+            )
+            if loc2glob:
+                ans = np.array(
+                    loc2glob(ans),
+                    dtype=float,
+                    order="C",
```

Comment:
is defining the `order` required? I think one can save a copy by using `np.column_stack` instead of `np.transpose`, if the order makes a big difference; probably worth making a note after benchmarking

Reply:
ok, I will change to np.column_stack